src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/CompletionListMerger.cs

```diff
@@ -55,6 +60,71 @@ internal static class CompletionListMerger
             return mergedCompletionList;
         }
 
+        public static object? MergeData(object? data1, object? data2)
+        {
+            if (data1 is null)
+            {
+                return data2;
+            }
+
+            if (data2 is null)
+            {
+                return data1;
+            }
+
+            return new MergedCompletionListData(data1, data2);
+        }
+
+        public static bool TrySplit(object? data, [NotNullWhen(true)] out IReadOnlyList<JObject>? splitData)
+        {
+            if (data is null)
+            {
+                splitData = null;
+                return false;
+            }
+
+            var collector = new List<JObject>();
+            Split(data, collector);
+
+            if (collector.Count == 0)
+            {
+                splitData = null;
+                return false;
+            }
+
+            splitData = collector;
+            return true;
+        }
+
+        private static void Split(object data, List<JObject> collector)
+        {
+            if (data is not JObject jobject)
+            {
+                return;
+            }
+
+            if (!jobject.ContainsKey(Data1Key) ||
+                !jobject.ContainsKey(Data2Key))
+            {
+                // Normal, non-merged data
+                collector.Add(jobject);
+            }
+            else
+            {
+                // Merged data
+                var mergedCompletionListData = jobject.ToObject<MergedCompletionListData>();
+
+                if (mergedCompletionListData is null)
+                {
+                    Debug.Fail("Merged completion list data is null, this should never happen.");
```

Comment:
Why check and then debug.fail, rather than just `Debug.Assert`? Seems odd.

Reply:
Debug.Assert wont result in us returning `null`. This is basically a "production safe" way to indicate that something is unexpected while simultaneously allowing the rest of the system to work as expected